drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,455 @@
+		}
+	}
+
+	return INVALID_CLIENT_IDX;
+}
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	/* Performing quick screening to see if it's worth doing more thorough
+	 * and cycle-demanding parsing of the received sring to determine if
+	 * it's a supported notification type.
+	 */
+	if ((len < AT_CMD_LEN(AT_CMD_CESQ_RESP)) ||
+	    (buf[4] != AT_CMD_CESQ_RESP[4])) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_RESP) ? true : false;
+}
+
+static bool is_cesq_subscribe(char *buf, size_t len)
+{
+	if (len < AT_CMD_LEN(AT_CMD_CESQ_ON)) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_ON) ? true : false;
+}
+
+static void cesq_notify(char *buf, size_t len)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(clients); i++) {
+		if (atomic_test_bit(&cesq_subscribers, i)) {
+			clients[i] = clients[i];
+			memcpy(clients[i]->response, buf, len);
+			clients[i]->response_len = len;
+
+			/* Dispatch handler through client's work. */
+			k_work_submit(&clients[i]->work);
+		}
+	}
+}
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	size_t client_idx;
+	struct at_cmd_client *client;
+	bool done;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	k_sem_take(&thread_start_sem, K_FOREVER);
+
+	for (;;) {
+		/* Workaround for current poll() implementation, that would
+		 * cause this loop to spin without yielding to lower priority
+		 * threads until data is received.
+		 */
+		k_sleep(10);
+
+		/* Poll the socket for incoming data. */
+		err = poll(poll_fds, nfds, K_FOREVER);
+		if (err < 0) {
+			LOG_ERR("Poll error: %d", err);
+			continue;
+		} else if (err == 0) {
+			LOG_DBG("Poll timed out.");
+			continue;
+		}
+
+		k_mutex_lock(&common_socket_mutex, K_FOREVER);
+
+		bytes_read = recv(common_socket_fd, buf,
+				  sizeof(buf), MSG_DONTWAIT);
+
+		k_mutex_unlock(&common_socket_mutex);
+
+		if (bytes_read <= 0) {
+			/* Current implementation of poll() will cause us to
+			 * get here many times before receiving data.
+			 */
+			continue;
+		}
+
+		if (is_cesq_notification(buf, bytes_read)) {
+			cesq_notify(buf, bytes_read);
+			continue;
+		}
+
+		/* If it's not a notification, find out which client has a
+		 * pending command, and dispatch the response to it.
+		 */
+		client_idx = client_cmd_pending_get_idx();
+		if (client_idx == INVALID_CLIENT_IDX) {
+			LOG_DBG("No client with pending command.");
+			continue;
+		}
+
+		client = clients[client_idx];
+
+		memcpy(client->response, buf, bytes_read);
+		client->response_len = bytes_read;
+
+		/* Dispatch handler through client's work. */
+		LOG_DBG("\t <- SUBMITTING WORK TO CLIENT %d", client->id);
+		k_work_submit(&client->work);
+
+		done = cmd_is_done(client->response, bytes_read);
+		if (done) {
+			k_poll_signal_raise(&cmd_done_signal, 1);
+			LOG_DBG("\t <- SIGNALED THAT COMMAND IS PROCESSED");
+		}
+	}
+}
+
+static void at_cmd_thread_fn(void *arg1, void *arg2, void *arg3)
```

Comment:
The reason it's done this way is because I don't want to start processing a new command from another (or the same) client before the previous one is finished. If I just offload this function as a work item to the workqueue, we risk waiting for quite a while for response from some commands, stalling the system workqueue. I tried to avoid spawning a separate thread, but didn't find a good solution. Will look more into it.

Reply:
Maybe you can use a mutex to manage that instead?